src/Symfony/Component/TypeInfo/TypeResolver/TypeResolver.php

```diff
@@ -0,0 +1,100 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\TypeInfo\TypeResolver;
+
+use PHPStan\PhpDocParser\Parser\PhpDocParser;
+use Psr\Container\ContainerInterface;
+use Symfony\Component\TypeInfo\Exception\UnsupportedException;
+use Symfony\Component\TypeInfo\Type;
+use Symfony\Component\TypeInfo\TypeContext\TypeContext;
+use Symfony\Component\TypeInfo\TypeContext\TypeContextFactory;
+
+/**
+ * Resolves type for a given subject by delegating resolving to nested type resolvers.
+ *
+ * @author Mathias Arlaud <mathias.arlaud@gmail.com>
+ * @author Baptiste Leduc <baptiste.leduc@gmail.com>
+ */
+final readonly class TypeResolver implements TypeResolverInterface
+{
+    /**
+     * @param ContainerInterface $resolvers Locator of type resolvers, keyed by supported subject type
+     */
+    public function __construct(
+        private ContainerInterface $resolvers,
+    ) {
+    }
+
+    public function resolve(mixed $subject, TypeContext $typeContext = null): Type
+    {
+        $subjectType = match (\is_object($subject)) {
+            true => match (true) {
+                is_subclass_of($subject::class, \ReflectionType::class) => \ReflectionType::class,
+                is_subclass_of($subject::class, \ReflectionFunctionAbstract::class) => \ReflectionFunctionAbstract::class,
+                default => $subject::class,
+            },
+            false => get_debug_type($subject),
+        };
```

Comment:
```suggestion
    $subjectType = \is_object($subject) ?
        match (true) {
            is_subclass_of($subject::class, \ReflectionType::class) => \ReflectionType::class,
            is_subclass_of($subject::class, \ReflectionFunctionAbstract::class) => \ReflectionFunctionAbstract::class,
            default => $subject::class,
        } : 
        get_debug_type($subject)
    ;
```

Reply:
I think the current version is more readable and I see no big improvement with your version of this snippet so I think I'll keep it as it is.